### PR TITLE
test: simplify test_runner.py

### PR DESCRIPTION
**Pull request by tdb3**
Implements the simplifications to test_runner.py proposed by sipa in PR #23995.

Remove the num_running variable as it can be implied by the length of the jobs list.

Remove the i variable as it can be implied by the length of the test_results list.

Instead of counting results to determine if finished, make the queue object itself 
responsible (by looking at running jobs and jobs left).

### CODE REVIEW

**Comment by kevkevinpal** [comment]
I tried this with `--failfast` and it seemed to properly fail fast

**Comment by tdb3** [comment]
> I tried this with `--failfast` and it seemed to properly fail fast

Thank you for the test

**Comment by kevkevinpal** [comment]
I tested with this diff e82b3421a8d5fb82fe20e670857abbbe220ae09e

and ran `./test/functional/test_runner.py --failfast`

File: test/functional/test_runner.py [hunk 1]
```diff
@@ -613,14 +613,12 @@ def run_tests(*, test_list, src_dir, build_dir, tmpdir, jobs=1, enable_coverage=
     max_len_name = len(max(test_list, key=len))
     test_count = len(test_list)
```
Comment (davidgumberg) [inline test/functional/test_runner.py:614]: Maybe we can remove `test_count`? See below.

File: test/functional/test_runner.py [hunk 2]
```diff
@@ -613,14 +613,12 @@ def run_tests(*, test_list, src_dir, build_dir, tmpdir, jobs=1, enable_coverage=
...
-    i = 0
-    while i < test_count:
+    while not job_queue.done():
         if failfast and not all_passed:
             break
         for test_result, testdir, stdout, stderr, skip_reason in job_queue.get_next():
             test_results.append(test_result)
-            i += 1
-            done_str = "{}/{} - {}{}{}".format(i, test_count, BOLD[1], test_result.name, BOLD[0])
+            done_str = "{}/{} - {}{}{}".format(len(test_results), test_count, BOLD[1], test_result.name, BOLD[0])
```
Comment (davidgumberg) [inline test/functional/test_runner.py:621]: Nit: We could use an f-string here and replace `test_count` with `len(test_list)`

```suggestion         
            done_str = f"{len(test_results)}/{len(test_list)} - {BOLD[1]}{test_result.name}{BOLD[0]}"

**Comment by tdb3** [comment]
Thanks for the review @daviddumberg!
Force pushed.  Responses inline.

File: test/functional/test_runner.py (hunk 1 above)
Comment (tdb3) [reply test/functional/test_runner.py:615]: `test_count` being assigned outside of the loop enables reporting of the correct total number of total tests on the right side of the `/`.

File: test/functional/test_runner.py (hunk 2 above)
Comment (tdb3) [reply test/functional/test_runner.py]: Thanks, updating to an f-string better aligns with the style guidelines (and that line is being changed anyway).
Leaving other format strings untouched for now (those could be updated in another refactor PR).

**Review by marcofleon — APPROVED** [review]
Tested ACK 8f2d814407c280f863be0303b4c2c1efdc0b8f36. 

Reviewed the code changes and I think they make the program a bit more readable. I ran `test_runner.py` and there were no issues. Also tested with `--failfast` and it did indeed fail fast at the expected test.

**Comment by davidgumberg** [comment]
reACK https://github.com/bitcoin/bitcoin/commit/8f2d814407c280f863be0303b4c2c1efdc0b8f36

**Review by mzumsande — COMMENTED** [review]
Code Review ACK 8f2d814407c280f863be0303b4c2c1efdc0b8f36

**Comment by tdb3** [comment]
> Tested ACK 8f2d814407c280f863be0303b4c2c1efdc0b8f36. 
> 
> Reviewed the code changes and I think they make the program a bit more readable. I ran `test_runner.py` and there were no issues. Also tested with `--failfast` and it did indeed fail fast at the expected test.

Thank you for the review @marcofleon!

**Comment by tdb3** [comment]
> Code Review ACK 8f2d814407c280f863be0303b4c2c1efdc0b8f36

Thank you for the review @mzumsande

**Comment by tdb3** [comment]
rebased

**Comment by mzumsande** [comment]
re-ACK 0831b54

>rebased

Why? This looked ready for merge with 3 ACKs. A rebase invalidates existing ACKs, so once these exist it should only be done if there is a conflict (Drahtbot will complain) or a specific reason (such as. a silent conflict). The CI will always rebase on master anyway for its runs (which could be restarted without rebasing if really necessary).

**Comment by tdb3** [comment]
> re-ACK 0831b54
> 
> >rebased
> 
> Why? This looked ready for merge with 3 ACKs. A rebase invalidates existing ACKs, so once these exist it should only be done if there is a conflict (Drahtbot will complain) or a specific reason (such as. a silent conflict). The CI will always rebase on master anyway for its runs (which could be restarted without rebasing if really necessary).

Thank you for the insight, didn't realize that. I'll avoid that moving forward. The rationale was to bring it to head while reinitiating CI (MacOS brew issue).

**Comment by davidgumberg** [comment]
reACK https://github.com/bitcoin/bitcoin/commit/0831b54dfca1b9e728295fff500215da14589fc0
Verified that the commit is unchanged with range-diff, and re-ran `make check`: 

```
$ git range-diff a945f09...8f2d814 1105aa4..0831b54
1:  8f2d814407 = 1:  0831b54dfc test: simplify test_runner.py
```

**Review by marcofleon — COMMENTED** [review]
re-ACK 0831b54dfca1b9e728295fff500215da14589fc0

**Review by maflcko — APPROVED** [review]
lgtm

**Pull request merged**